File: wpilibjExamples/src/main/java/edu/wpi/first/wpilibj/examples/trajectorycommand/subsystems/DriveSubsystem.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.wpilibj.examples.trajectorycommand.subsystems;

import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.math.controller.RamseteController;
import edu.wpi.first.math.controller.SimpleMotorFeedforward;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.math.kinematics.DifferentialDriveOdometry;
import edu.wpi.first.math.kinematics.DifferentialDriveWheelSpeeds;
import edu.wpi.first.math.trajectory.Trajectory;
import edu.wpi.first.wpilibj.ADXRS450_Gyro;
import edu.wpi.first.wpilibj.Encoder;
import edu.wpi.first.wpilibj.drive.DifferentialDrive;
import edu.wpi.first.wpilibj.examples.trajectorycommand.Constants.AutoConstants;
import edu.wpi.first.wpilibj.examples.trajectorycommand.Constants.DriveConstants;
import edu.wpi.first.wpilibj.interfaces.Gyro;
import edu.wpi.first.wpilibj.motorcontrol.MotorControllerGroup;
import edu.wpi.first.wpilibj.motorcontrol.PWMSparkMax;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import edu.wpi.first.wpilibj2.command.TrajectoryCommand;

public class DriveSubsystem extends SubsystemBase {
  // The motors on the left side of the drive.
  private final MotorControllerGroup m_leftMotors =
      new MotorControllerGroup(
          new PWMSparkMax(DriveConstants.kLeftMotor1Port),
          new PWMSparkMax(DriveConstants.kLeftMotor2Port));

  // The motors on the right side of the drive.
  private final MotorControllerGroup m_rightMotors =
      new MotorControllerGroup(
          new PWMSparkMax(DriveConstants.kRightMotor1Port),
          new PWMSparkMax(DriveConstants.kRightMotor2Port));

  // The robot's drive
  private final DifferentialDrive m_drive = new DifferentialDrive(m_leftMotors, m_rightMotors);

  // The left-side drive encoder
  private final Encoder m_leftEncoder =
      new Encoder(
          DriveConstants.kLeftEncoderPorts[0],
          DriveConstants.kLeftEncoderPorts[1],
          DriveConstants.kLeftEncoderReversed);

  // The right-side drive encoder
  private final Encoder m_rightEncoder =
      new Encoder(
          DriveConstants.kRightEncoderPorts[0],
          DriveConstants.kRightEncoderPorts[1],
          DriveConstants.kRightEncoderReversed);

  // The gyro sensor
  private final Gyro m_gyro = new ADXRS450_Gyro();

  // Odometry class for tracking robot pose
  private final DifferentialDriveOdometry m_odometry;
  private final RamseteController m_ramseteController =
      new RamseteController(AutoConstants.kRamseteB, AutoConstants.kRamseteZeta);
  private final SimpleMotorFeedforward m_feedforward =
      new SimpleMotorFeedforward(
          DriveConstants.ksVolts,
          DriveConstants.kvVoltSecondsPerMeter,
          DriveConstants.kaVoltSecondsSquaredPerMeter);
  private final PIDController m_leftController = new PIDController(DriveConstants.kPDriveVel, 0, 0);
  private final PIDController m_rightController =
      new PIDController(DriveConstants.kPDriveVel, 0, 0);

  // Track previous target velocities for feedforward calculation
  private DifferentialDriveWheelSpeeds m_previousSpeeds = new DifferentialDriveWheelSpeeds();

  /** Creates a new DriveSubsystem. */
  public DriveSubsystem() {
    // Sets the distance per pulse for the encoders
    m_leftEncoder.setDistancePerPulse(DriveConstants.kEncoderDistancePerPulse);
    m_rightEncoder.setDistancePerPulse(DriveConstants.kEncoderDistancePerPulse);

    resetEncoders();
    m_odometry = new DifferentialDriveOdometry(m_gyro.getRotation2d());
  }

  @Override
  public void periodic() {
    // Update the odometry in the periodic block
    m_odometry.update(
        m_gyro.getRotation2d(), m_leftEncoder.getDistance(), m_rightEncoder.getDistance());
  }

  /**
   * Returns the currently-estimated pose of the robot.
   *
   * @return The pose.
   */
  public Pose2d getPose() {
    return m_odometry.getPoseMeters();
  }

  /**
   * Resets the odometry to the specified pose.
   *
   * @param pose The pose to which to set the odometry.
   */
  public void resetOdometry(Pose2d pose) {
    resetEncoders();
    m_odometry.resetPosition(pose, m_gyro.getRotation2d());
  }

  /**
   * Drives the robot using arcade controls.
   *
   * @param fwd The commanded forward movement
   * @param rot The commanded rotation
   */
  public void arcadeDrive(double fwd, double rot) {
    m_drive.arcadeDrive(fwd, rot);
  }

  /** Resets the drive encoders to currently read a position of 0. */
  public void resetEncoders() {
    m_leftEncoder.reset();
    m_rightEncoder.reset();
  }

  /**
   * Gets the average distance of the two encoders.
   *
   * @return The average of the two encoder readings
   */
  public double getAverageEncoderDistance() {
    return (m_leftEncoder.getDistance() + m_rightEncoder.getDistance()) / 2.0;
  }

  /**
   * Sets the max output of the drive. Useful for scaling the drive to drive more slowly.
   *
   * @param maxOutput The maximum output to which the drive will be constrained
   */
  public void setMaxOutput(double maxOutput) {
    m_drive.setMaxOutput(maxOutput);
  }

  /** Zeroes the heading of the robot. */
  public void zeroHeading() {
    m_gyro.reset();
  }

  /**
   * Returns the heading of the robot.
   *
   * @return The robot's heading in degrees, from -180 to 180
   */
  public double getHeading() {
    return m_gyro.getRotation2d().getDegrees();
  }

  /**
   * Returns the turn rate of the robot.
   *
   * @return The turn rate of the robot, in degrees per second
   */
  public double getTurnRate() {
    return -m_gyro.getRate();
  }

  /**
   * Reaches the given target state. Used internally to follow the TrajectoryCommand's output.
   *
   * @param targetState The target state struct.
   */
  private void followState(Trajectory.State targetState) {
    // Calculate via Ramsete the speed vector needed to reach the target state
    ChassisSpeeds speedVector =
        m_ramseteController.calculate(m_odometry.getPoseMeters(), targetState);
    // Convert the vector to the separate velocities of each side of the drivetrain
    DifferentialDriveWheelSpeeds targetSpeeds =
        DriveConstants.kDriveKinematics.toWheelSpeeds(speedVector);

    // PID and Feedforward control
    // This can be replaced by calls to smart motor controller closed-loop control
    // functionality. For example:
    // mySmartMotorController.setSetpoint(leftMetersPerSecond, ControlType.Velocity);

    double leftFeedforward =
        m_feedforward.calculate(
            m_previousSpeeds.leftMetersPerSecond,
            targetSpeeds.leftMetersPerSecond,
            AutoConstants.kTimestepSeconds);

    double rightFeedforward =
        m_feedforward.calculate(
            m_previousSpeeds.rightMetersPerSecond,
            targetSpeeds.rightMetersPerSecond,
            AutoConstants.kTimestepSeconds);

    double leftOutput =
        leftFeedforward
            + m_leftController.calculate(m_leftEncoder.getRate(), targetSpeeds.leftMetersPerSecond);

    double rightOutput =
        rightFeedforward
            + m_rightController.calculate(
                m_rightEncoder.getRate(), targetSpeeds.rightMetersPerSecond);

    // Apply the voltages
    m_leftMotors.setVoltage(leftOutput);
    m_rightMotors.setVoltage(rightOutput);
    m_drive.feed();

    // Track previous speed setpoints
    m_previousSpeeds = targetSpeeds;
  }

  /**
   * Build a command group for the given trajectory.
   *
   * <p>The group consists of:
   *
   * <p>- resetting the odometry to the trajectory's initial pose
   *
   * <p>- following the trajectory
   *
   * <p>- stopping at the end of the path
   *
   * @param trajectory The path to follow
   * @return A command group that tracks the given trajectory
   */
  public Command buildTrajectoryGroup(Trajectory trajectory) {
    return new TrajectoryCommand(trajectory, this::followState, this)
        // Reset odometry to the starting pose of the trajectory.
        .beforeStarting(() -> resetOdometry(trajectory.getInitialPose()), this)
        // Stop at the end of the trajectory.
        .andThen(
            () -> {
              m_previousSpeeds = new DifferentialDriveWheelSpeeds();
              m_drive.stopMotor();
            },
            this);
  }
}

File: wpilibcExamples/src/main/cpp/examples/TrajectoryCommand/cpp/subsystems/DriveSubsystem.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "subsystems/DriveSubsystem.h"

#include <frc/geometry/Rotation2d.h>
#include <frc/kinematics/ChassisSpeeds.h>
#include <frc/kinematics/DifferentialDriveWheelSpeeds.h>
#include <frc2/command/InstantCommand.h>
#include <frc2/command/SequentialCommandGroup.h>
#include <frc2/command/TrajectoryCommand.h>
#include <units/voltage.h>

using namespace DriveConstants;
using namespace AutoConstants;

DriveSubsystem::DriveSubsystem()
    : m_left1{kLeftMotor1Port},
      m_left2{kLeftMotor2Port},
      m_right1{kRightMotor1Port},
      m_right2{kRightMotor2Port},
      m_leftEncoder{kLeftEncoderPorts[0], kLeftEncoderPorts[1]},
      m_rightEncoder{kRightEncoderPorts[0], kRightEncoderPorts[1]},
      m_ramseteController{kRamseteB, kRamseteZeta},
      m_leftController{kPDriveVel, 0, 0},
      m_rightController{kPDriveVel, 0, 0},
      m_feedforward{kS, kV, kA},
      m_odometry{m_gyro.GetRotation2d()} {
  // Set the distance per pulse for the encoders
  m_leftEncoder.SetDistancePerPulse(kEncoderDistancePerPulse);
  m_rightEncoder.SetDistancePerPulse(kEncoderDistancePerPulse);

  ResetEncoders();
}

void DriveSubsystem::Periodic() {
  // Implementation of subsystem periodic method goes here.
  m_odometry.Update(m_gyro.GetRotation2d(),
                    units::meter_t(m_leftEncoder.GetDistance()),
                    units::meter_t(m_rightEncoder.GetDistance()));
}

void DriveSubsystem::ArcadeDrive(double fwd, double rot) {
  m_drive.ArcadeDrive(fwd, rot);
}

void DriveSubsystem::FollowState(const frc::Trajectory::State& targetState) {
  // Calculate via Ramsete the speed vector needed to reach the target state
  frc::ChassisSpeeds speedVector =
      m_ramseteController.Calculate(m_odometry.GetPose(), targetState);
  // Convert the vector to the separate velocities of each side of the
  // drivetrain
  frc::DifferentialDriveWheelSpeeds targetSpeeds =
      kDriveKinematics.ToWheelSpeeds(speedVector);

  // PID and Feedforward control
  // This can be replaced by calls to smart motor controller closed-loop control
  // functionality. For example:
  // mySmartMotorController.setSetpoint(leftMetersPerSecond,
  // ControlType.Velocity);

  auto leftFeedforward = m_feedforward.Calculate(
      m_previousSpeeds.left, targetSpeeds.left, AutoConstants::kTimestep);

  auto rightFeedforward = m_feedforward.Calculate(
      m_previousSpeeds.right, targetSpeeds.right, AutoConstants::kTimestep);

  auto leftOutput = leftFeedforward + units::volt_t{m_leftController.Calculate(
                                          m_leftEncoder.GetRate(),
                                          targetSpeeds.left.to<double>())};

  auto rightOutput =
      rightFeedforward +
      units::volt_t{m_rightController.Calculate(
          m_rightEncoder.GetRate(), targetSpeeds.right.to<double>())};

  // Apply the voltages
  m_leftMotors.SetVoltage(leftOutput);
  m_rightMotors.SetVoltage(rightOutput);
  m_drive.Feed();

  // Track previous speed setpoints
  m_previousSpeeds = targetSpeeds;
}

void DriveSubsystem::ResetEncoders() {
  m_leftEncoder.Reset();
  m_rightEncoder.Reset();
}

double DriveSubsystem::GetAverageEncoderDistance() {
  return (m_leftEncoder.GetDistance() + m_rightEncoder.GetDistance()) / 2.0;
}

void DriveSubsystem::SetMaxOutput(double maxOutput) {
  m_drive.SetMaxOutput(maxOutput);
}

units::degree_t DriveSubsystem::GetHeading() const {
  return m_gyro.GetRotation2d().Degrees();
}

double DriveSubsystem::GetTurnRate() {
  return -m_gyro.GetRate();
}

frc::Pose2d DriveSubsystem::GetPose() {
  return m_odometry.GetPose();
}

void DriveSubsystem::ResetOdometry(frc::Pose2d pose) {
  ResetEncoders();
  m_odometry.ResetPosition(pose, m_gyro.GetRotation2d());
}

frc2::Command* DriveSubsystem::BuildTrajectoryGroup(
    const frc::Trajectory& trajectory) {
  return new frc2::SequentialCommandGroup(
      frc2::InstantCommand(
          [this, trajectory] { ResetOdometry(trajectory.InitialPose()); }),
      frc2::TrajectoryCommand(
          trajectory,
          [this](const frc::Trajectory::State& targetState) {
            FollowState(targetState);
          },
          {this}),
      frc2::InstantCommand([this] {
        m_drive.StopMotor();
        m_previousSpeeds = frc::DifferentialDriveWheelSpeeds();
      }));
}
